pkg/sql/idxrecommendations/idx_recommendations_cache.go

```diff
@@ -0,0 +1,199 @@
+		// If couldn't initialized, don't generate recommendations.
+		if !found {
+			return false
+		}
+	}
+
+	timeSinceLastGenerated := timeutil.Now().Sub(recInfo.lastGeneratedTs)
+	return recInfo.executionCount >= 5 && timeSinceLastGenerated.Hours() >= 1
+}
+
+// UpdateIndexRecommendations updates the values for index recommendations.
+// If reset is true, a new recommendation was generated, so reset the execution counter and
+// lastGeneratedTs, otherwise just increment the executionCount.
+func (idxRec *IndexRecCache) UpdateIndexRecommendations(
+	fingerprint string, planHash uint64, database string, recommendations []string, reset bool,
+) []string {
+	idxKey := indexRecKey{
+		stmtNoConstants: fingerprint,
+		database:        database,
+		planHash:        planHash,
+	}
+
+	if reset {
+		idxRec.setIndexRecommendations(idxKey, timeutil.Now(), recommendations, 0)
+		return recommendations
+	}
+
+	recInfo, found := idxRec.getIndexRecommendation(idxKey)
+	// If is the first time, we want the lastGeneratedTs to be in the past, in case we reach
+	// the execution count, we should generate new recommendations.
+	generatedTime := timeutil.Now().Add(-duration.SecsPerDay * 1e6)
+	count := int64(1)
+	if found {
+		generatedTime = recInfo.lastGeneratedTs
+		recommendations = recInfo.recommendations
+		count = recInfo.executionCount + 1
+	}
+	idxRec.setIndexRecommendations(idxKey, generatedTime, recommendations, count)
+
+	return recInfo.recommendations
+}
+
+func (idxRec *IndexRecCache) getIndexRecommendation(key indexRecKey) (indexRecInfo, bool) {
+	idxRec.mu.RLock()
+	defer idxRec.mu.RUnlock()
+
+	recInfo, found := idxRec.mu.idxRecommendations[key]
+
+	return recInfo, found
+}
+
+func (idxRec *IndexRecCache) setIndexRecommendations(
+	key indexRecKey, time time.Time, recommendations []string, execCount int64,
+) {
+	idxRec.mu.RLock()
+	defer idxRec.mu.RUnlock()
+
+	_, found := idxRec.getIndexRecommendation(key)
+
+	// If it was not found, check if a new entry can be created, without
+	// passing the limit of unique index recommendations from the cache.
+	if !found {
+		limit := idxRec.uniqueIndexRecInfoLimit.Get(&idxRec.st.SV)
+		incrementedCount :=
+			atomic.AddInt64(idxRec.atomic.uniqueIndexRecInfo, int64(1))
+
+		// If we have exceeded limit of unique index recommendations try to delete older data.
+		if incrementedCount > limit {
+			deleted := idxRec.ClearOlderIndexRecommendationsCache()
+			// Abort if no entries were deleted.
+			if deleted == 0 {
+				atomic.AddInt64(idxRec.atomic.uniqueIndexRecInfo, -int64(1))
+				return
+			}
+		}
+	}
+
+	idxRec.mu.idxRecommendations[key] = indexRecInfo{
+		lastGeneratedTs: time,
+		recommendations: recommendations,
+		executionCount:  execCount,
+	}
+}
+
+// ClearOlderIndexRecommendationsCache clear entries that was last updated
+// more than a day ago. Returns the total deleted entries.
+func (idxRec *IndexRecCache) ClearOlderIndexRecommendationsCache() int {
```

Comment:
This is a very wordy method name. Also, does is need to be exported?

Reply:
Initially I was considering also calling this function in another location periodically to help with the cleanup, but decided to keep it just here. Renamed to be local and smaller